### ISSUE TITLE
matrix multiplication @ vs. `np.dot`

### ISSUE THREAD

**Issue by ReHoss** [microsoft/pyright#8526]
Probably the expected behaviour. However, could you explain why such error is raised?

```python
import numpy as np
import numpy.typing as npt


def test():
    a = np.eye(2, dtype="float32")
    b = np.eye(2, dtype="float32")

    c: npt.NDArray[np.floating] = a @ b
    d: npt.NDArray[np.floating] = np.dot(a, b)

    del a, b, c, d
```

why does `a @ b` returns

```
error: Expression of type "NDArray[bool_]" is incompatible with declared type "NDArray[floating[Unknown]]"
    "ndarray[Any, dtype[bool_]]" is incompatible with "ndarray[Any, dtype[floating[Unknown]]]"
      Type parameter "_DType_co@ndarray" is covariant, but "dtype[bool_]" is not a subtype of "dtype[floating[Unknown]]"
        "dtype[bool_]" is incompatible with "dtype[floating[Unknown]]"
          Type parameter "_DTypeScalar_co@dtype" is covariant, but "bool_" is not a subtype of "floating[Unknown]"
            "bool_" is incompatible with "floating[Unknown]" (reportAssignmentType)
1 error, 0 warnings, 0 informations 
```

**Comment by erictraut** [comment]
This is related to the fact that numpy's type stubs include some complex overloads that rely on behaviors that are underspecified in the Python typing spec. When pyright evaluates `a @ b`, it determines that the type is `NDArray[bool_]` which is not assignable to `NDArray[floating]`.

Mypy doesn't generate an error here because its overload evaluation behavior is subtly different from pyright's. It evaluates the type of `a @ b` as `Any`.

I would like to see the type evaluation rules for overloads be fully and unambiguously described in the typing spec. If and when that's done, I will update pyright to conform to the spec. We've been making [good progress](https://docs.google.com/presentation/d/1ash1ellodiV6FTyVnY-Q6V2G5vJ68ycWBr6MYkayC1U/edit#slide=id.g2770f948faa_0_0) in clarifying other parts of the typing spec, but this particular area hasn't been tackled yet. It's likely to be controversial, and it may take some time to get these rules pinned down, but I think it will be worth it in the end if we can collectively agree on a standard behavior.

I don't plan to make any changes to pyright's overload resolution behavior until the spec is updated because doing so will likely just create churn for pyright users (resulting in new false positives and negatives within existing typed code bases). 

I'm going to close the issue since pyright is working as intended here, so I don't consider this a bug.

**Comment by ReHoss** [comment]
@erictraut, thank you
